sympy/integrals/transforms.py

```diff
@@ -1005,14 +1007,37 @@ def repl(ex, *args):
     expr = repl(expr, Unequality, replue)
     return S(expr)
 
+def expand_dirac_delta(expr):
+    """
+    Expand an expression involving DiractDelta to get it as a linear
+    combination of DiracDelta functions.
+    """
+    return _lin_eq2dict(expr, expr.atoms(DiracDelta))
 
 @_noconds
 def _laplace_transform(f, t, s_, simplify=True):
     """ The backend function for Laplace transforms. """
     from sympy import (re, Max, exp, pi, Min, periodic_argument as arg_,
-                       arg, cos, Wild, symbols, polar_lift)
+                       arg, cos, Wild, symbols, polar_lift, diff)
     s = Dummy('s')
-    F = integrate(exp(-s*t) * f, (t, 0, oo))
+    if f.has(DiracDelta):
+        try:
+            nodelta, deltadict = expand_dirac_delta(f)
+        except PolyNonlinearError:
+            raise IntegralTransformError(
+            'Laplace', f, 'could not expand DiracDelta expressions')
+        _parts = [ integrate(exp(-s*t) * nodelta, (t, 0, oo)) ]
+        for dd in deltadict.keys():
+            if len(dd.args)>1 and dd.args[1]>0:
+                raise IntegralTransformError(
+                    'Laplace', f, 'derivatives of DiracDelta not implemented')
+            if dd.args[0].subs(t,0) == 0:
+                _parts.append(deltadict[dd].subs(t,0)/diff(dd.args[0],t).subs(t,0))
+            else:
+                _parts.append(integrate(exp(-s*t)*dd*deltadict[dd],(t, 0, oo)))
+        F = sum(_parts)
+    else:
+        F = integrate(exp(-s*t) * f, (t, 0, oo))
```

Comment:
I think this can be simplified.

We don't need to check `has(DiracDelta)` since `expand_dirac_delta` can handle the case where there is no `DiracDelta`.

We should only call `integrate` once. I think actually the whole expression should be integrated. We only need to extract those DiracDelta terms that would contribute to the result but that would be excluded by `integrate`. Those terms can be added to the result of integrating the whole expression. This is also avoid problems where the argument to delta has multiple zeros e.g. `DiracDelta(x*(1 - x))`.

It's better to use `Add(*terms)` rather than `sum(terms)` because `sum` is `O(n^2)` when used with sympy expressions.

I think that we also need to check for the case that the derivative of the DiracDelta argument is zero.

Checking with `==0` is unreliable. Instead `is_zero` should be used and the three separate cases True, False and None should be handled. In the None case either `IntegralTransformError` could be raised or a Piecewise result could be constructed.

Reply:
> I think this can be simplified.

Thank you! All of it makes a lot of sense; I will do it next week. One thing I am not sure about:

> I think that we also need to check for the case that the derivative of the DiracDelta argument is zero.

You mean like `DiracDelta(a*x**2)`? So what is the result in that case? Would it always go towards $\infty$, or would it depend on higher derivatives or on the sign of the argument?

Reply:
P.S. I already found time today, it works well with all examples.

I have done everything except the `slope=0` thing. In case you already want to look at what I have, I push it now.